liquibase-core/src/main/java/liquibase/integration/spring/SpringLiquibase.java

```diff
@@ -340,47 +342,58 @@ private void generateRollbackFile(Liquibase liquibase) throws LiquibaseException
     }
 
     protected void performUpdate(Liquibase liquibase) throws LiquibaseException {
-		if (isTestRollbackOnUpdate()) {
-			if (tag != null) {
-				liquibase.updateTestingRollback(tag, new Contexts(getContexts()), new LabelExpression(getLabels()));
-			} else {
-				liquibase.updateTestingRollback(new Contexts(getContexts()), new LabelExpression(getLabels()));
-			}
-		} else {
-			if (tag != null) {
-				liquibase.update(tag, new Contexts(getContexts()), new LabelExpression(getLabels()));
-			} else {
-				liquibase.update(new Contexts(getContexts()), new LabelExpression(getLabels()));
-			}
-		}
-    }
-
-	protected Liquibase createLiquibase(Connection c) throws LiquibaseException {
-		SpringResourceOpener resourceAccessor = createResourceOpener();
-		Liquibase liquibase = new Liquibase(getChangeLog(), resourceAccessor, createDatabase(c, resourceAccessor));
+        if (isTestRollbackOnUpdate()) {
+            if (tag != null) {
+                liquibase.updateTestingRollback(tag, new Contexts(getContexts()), new LabelExpression(getLabels()));
+            } else {
+                liquibase.updateTestingRollback(new Contexts(getContexts()), new LabelExpression(getLabels()));
+            }
+        } else {
+            if (tag != null) {
+                liquibase.update(tag, new Contexts(getContexts()), new LabelExpression(getLabels()));
+            } else {
+                liquibase.update(new Contexts(getContexts()), new LabelExpression(getLabels()));
+            }
+        }
+    }
+
+    protected void performUpdateIfNeeded(final Liquibase liquibase) throws LiquibaseException {
+        if (updateNeeded(liquibase, new Contexts(getContexts()), new LabelExpression(getLabels()))) {
```

Comment:
@dancek thanks for this solution! We tried to apply it and find out that adding 

`ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(liquibase.getDatabase()).reset();`

 before performUpdate(liquibase);  helps to avoid unnecessary DB update in case of multiple server nodes simultaneous startup.